### PR TITLE
Detach JavaMail

### PR DESCRIPTION
**Pull request by basil**
### Background

See [the mailing list thread](https://groups.google.com/g/jenkinsci-dev/c/kZGpxBVq6Sk/m/Jd2ssj7xDwAJ).

### Problem

Core currently ships JavaMail and JavaBeans Activation Framework (JAF), a dependency of JavaMail. Core only consumes this functionality in a tiny validation routine (`JenkinsLocationConfiguration#doCheckAdminAddress`), but many plugins consume this functionality via core. Both libraries are subject to Jakarta package renaming changes. To comply with these changes, a migration plan is needed.

### Solution

We have released four new plugins:

- [`javax-activation-api`](https://plugins.jenkins.io/javax-activation-api), bundling `javax.activation`
- [`javax-mail-api`](https://plugins.jenkins.io/javax-mail-api), bundling `javax.mail` and depending on `javax-activation-api`
- [`jakarta-activation-api`](https://plugins.jenkins.io/jakarta-activation-api), bundling `jakarta.activation`
- [`jakarta-mail-api`](https://plugins.jenkins.io/jakarta-mail-api), bundling `jakarta.mail` and depending on `jakarta-activation-api`

In this PR we detach the JavaMail and JAF JARs from core to `javax-mail-api` and `javax-activation-api`. Existing plugins that depend on the `javax` packages will continue to work. Upon upgrading their core baseline, these plugins can retain their use of Java EE (by adding an explicit dependency on `javax-mail-api` and/or `javax-activation-api`) or migrate to Jakarta EE (by adding an explicit dependency on `jakarta-mail-api` and/or `jakarta-activation-api`) as desired.

### Implementation

The reason why we add a new test dependency on `javax-mail-api` is because we have an existing test dependency on `mailer`. In a real-world scenario, installing the current version of `mailer` against a fresh Jenkins home would also result in installing `javax-mail-api`. But the test framework doesn't load plugins the same way as in production and we get warnings that cause `jenkins.model.StartupTest#noWarnings` to fail. Since it is not possible to end up in this situation in production, I have opted to make the test scenario closer to the production scenario by ensuring that the test always runs with both `mailer` and `javax-mail-api` installed. This gets the test to pass and also matches the production scenario where `mailer` cannot be installed without also installing `javax-mail-api`.

### Testing Done

Started Jenkins with `java -jar jenkins.war` against an older home directory and observed the following logs:

```
INFO    hudson.PluginManager#loadDetachedPlugins: Upgrading Jenkins. The last running version was 2.327-SNAPSHOT. This Jenkins is version 2.329-SNAPSHOT.
INFO    hudson.PluginManager#loadDetachedPlugins: Upgraded Jenkins from version 2.327-SNAPSHOT to version 2.329-SNAPSHOT. Loaded detached plugins (and dependencies): [javax-mail-api.hpi, javax-activation-api.hpi]
```

Started Jenkins with `java-jar jenkins.war` against a fresh home directory, installed the suggested plugin set, and observed the following logs:

```
INFO    hudson.PluginManager#install: Starting installation of a batch of 19 plugins plus their dependencies
INFO    hudson.model.UpdateSite$Plugin#deploy: Adding dependent install of javax-mail-api for plugin cloudbees-folder
INFO    hudson.model.UpdateSite$Plugin#deploy: Adding dependent install of javax-activation-api for plugin javax-mail-api
```

Ensured that the contents of the WAR were identical to the contents before this PR modulo the removal of the JavaMail and JAF JARs.

### Proposed changelog entries

The [JavaMail](https://github.com/javaee/javamail) and [JavaBeans Activation Framework (JAF)](https://github.com/javaee/activation) libraries have been detached from Jenkins core into dedicated plugins.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/jenkins/model/JenkinsLocationConfiguration.java
Comment (timja) [inline core/src/main/java/jenkins/model/JenkinsLocationConfiguration.java:213]: could we just replace this whole method with something like:

```java
Pattern.compile("^(.+)@(\\S+)$")
      .matcher(value)
      .matches()
```

(obviously can make the pattern a constant)

from https://www.baeldung.com/java-email-validation-regex

Comment (basil) [reply core/src/main/java/jenkins/model/JenkinsLocationConfiguration.java:213]: Possibly. Feel free to propose a separate PR doing so if desired. This PR is maintaining the existing status quo.

Comment (timja) [reply core/src/main/java/jenkins/model/JenkinsLocationConfiguration.java:213]: =/ I don't think that adding 30 lines of reflection is 'maintaining' the status quo.

If we're removing a dependency then we either should have a simple way of validating it or removing the validation.

Comment (basil) [reply core/src/main/java/jenkins/model/JenkinsLocationConfiguration.java:213]: It _is_ maintaining the status quo in two senses:

- Before, we called `InternetAddress(String)` to do the validation. This PR continues to call the same validation routine.
- Before, when detaching functionality we have used reflection to maintain anti-modular interfaces (e.g., `AbstractBuild#getTestResultAction` and `AbstractBuild#getAggregatedTestResultAction`. This PR follows the same precedent.

I am not going to change the status quo in this PR. If you feel that this implementation should be replaced with a regular expression or deleted, feel free to propose a separate PR for that. If and when it is accepted, I will merge it with this PR (if this PR isn't shipped first).

Comment (jglick) [reply core/src/main/java/jenkins/model/JenkinsLocationConfiguration.java:213]: > when detaching functionality we have used reflection to maintain anti-modular interfaces (e.g., `AbstractBuild#getTestResultAction` and `AbstractBuild#getAggregatedTestResultAction`)

Those cases were unfortunate but required to prevent serious regression of basic CI functionality; there was just no way around it. This bit of form validation (on what I am guessing is a rather unimportant setting—in core, seems to be used only for RSS feeds) does not in my opinion justify use of reflection.

Comment (basil) [reply core/src/main/java/jenkins/model/JenkinsLocationConfiguration.java:213]: In my opinion it is good enough, and it is all I am willing to offer. Take it or leave it. If you don't want this PR, feel free to close it.

Comment (jglick) [reply core/src/main/java/jenkins/model/JenkinsLocationConfiguration.java:213]: > feel free to propose a separate PR for that

in progress

Comment (basil) [reply core/src/main/java/jenkins/model/JenkinsLocationConfiguration.java]: Rebased on top of #6166

File: bom/pom.xml
Comment (jglick) [inline bom/pom.xml:325]: Since this confused me: `WEB-INF/lib/jakarta.mail-1.6.5.jar` actually contains `javax.mail` (in addition to `com.sun.mail`) despite the artifact name. https://eclipse-ee4j.github.io/mail/docs/JavaMail-2.0-changes.txt explains

> Previous version of the Jakarta Mail Specification which is based on
> JavaMail 1.6 used javax.mail package namespace.
> New Jakarta Mail Specification uses jakarta.mail package namespace instead.

File: war/pom.xml
Comment (jglick) [inline war/pom.xml:379]: We used to bundle 1.6.5 and now we are detaching 1.6.2? https://github.com/jenkinsci/javax-mail-api-plugin/blob/javax-mail-api-1.6.2-1/pom.xml#L47 vs. https://github.com/jenkinsci/jenkins/blob/c31d9302d946463884406f920c8d14b2cf7cdf24/bom/pom.xml#L325

BTW you may consider https://www.jenkins.io/doc/developer/publishing/releasing-cd/#pom-file-modifications having plugin version automatically track library version. Otherwise it is all too easy for Dependabot to let https://github.com/jenkinsci/javax-mail-api-plugin/blob/e8b4d80f40e47e76aad413366953d333b0975f97/pom.xml#L36-L47 get out of synch.

File: bom/pom.xml
Comment (basil) [reply bom/pom.xml:325]: Correct

File: war/pom.xml
Comment (basil) [reply war/pom.xml:379]: > We used to bundle 1.6.5 and now we are detaching 1.6.2?

Correct. Really the root cause of this problem is that the Jakarta organization published a few releases under the original package names before switching to the Jakarta package names. I suppose we _could_ release a 1.6.5 version of `javax-mail-api` that switches to the last Jakarta organization release prior to the Jakarta package rename. But then plugins wouldn't be able to depend on `javax-mail-api` and `jakarta-mail-api` concurrently without awkward Enforcer conflicts. Seemed simpler to avoid the issue by sticking to the last pre-Jakarta-organization release. I have not been able to find any meaningful differences between the last "pre-Jakarta-organization" release and the last "post-Jakarta-organization-change-pre-package-rename" release that would affect end users, much less any binary incompatibilities.

**Review by timja — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
Would be nice to have downstream PRs to adjust former changes like https://github.com/jenkinsci/git-plugin/pull/1101 in Git Plugin. 

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by timja** [comment]
@basil did you want to do a bom run before merging this? or happy without? (non blocking)

**Comment by basil** [comment]
Sure, I'll do a BOM test once I get a new incremental.

File: war/pom.xml
Comment (basil) [reply war/pom.xml:379]: It's a bit tough to explain, but 1.6.2 was the last release under the original `groupId` and `artifactId` (and used the `javax.` package namespace). 1.6.3, 1.6.4, and 1.6.5 were released under a new Jakarta `groupId` and `artifactId` (but still used the `javax.` package namespace). 2.0.0 was released under the same Jakarta `groupId` and `artifactId` (but used the `jakarta.` package namespace). So this makes things a bit awkward if you want to put both the old `javax.` classes and the new `jakarta.` packages on the same classpath. You can't easily do that with 1.6.3, 1.6.4, and 1.6.5 because they share the same `groupId` and `artifactId` as 2.0.0 (despite having different packages).

This doesn't matter much for production, but it does matter for plugin/BOM tests that might pull in e.g. both `javax-mail-api` and `jakarta-mail-api` as test-scoped dependencies. If I had used 1.6.5 in `javax-mail-api` that would have forced me to use the Jakarta `groupId` and `artifactId`, which would then conflict with the 2.0 version from `jakarta-mail-api`. If you want to have both sets of packages on the classpath at the same time, you basically need to stick to the original pre-Jakarta group ID and artifact ID, which also means being behind a few very small minor releases.

Just how small? See https://eclipse-ee4j.github.io/mail/docs/CHANGES.txt

Comment (jglick) [reply war/pom.xml:379]: OK, makes sense now, thanks for explanation!

**Comment by basil** [comment]
> Sure, I'll do a BOM test once I get a new incremental.

Actually BOM testing of all core incrementals is currently broken due to jenkinsci/incrementals-tools#28, but I will do a BOM test once that is fixed.

**Comment by basil** [comment]
[BOM test passed](https://ci.jenkins.io/job/Tools/job/bom/job/PR-825/1/)

**Comment by basil** [comment]
> Would be nice to have downstream PRs

I have created the following downstream PRs:

- jenkinsci/mailer-plugin#142
- jenkinsci/mailer-plugin#143
- jenkinsci/mailer-plugin#144
- jenkinsci/email-ext-plugin#335
- jenkinsci/email-ext-plugin#336
- jenkinsci/email-ext-plugin#337
- jenkinsci/workflow-basic-steps-plugin#166
- jenkinsci/workflow-basic-steps-plugin#167
- jenkinsci/workflow-basic-steps-plugin#168

**Comment by basil** [comment]
End-to-end testing failed with:

```
DEBUG SMTP: IOException while sending, closing, THROW: 
javax.activation.UnsupportedDataTypeException: no object DCH for MIME type multipart/mixed; 
	boundary="----=_Part_5_726582541.1641663022150"
	at javax.activation.ObjectDataContentHandler.writeTo(DataHandler.java:909)
	at javax.activation.DataHandler.writeTo(DataHandler.java:330)
	at javax.mail.internet.MimeBodyPart.writeTo(MimeBodyPart.java:1694)
	at javax.mail.internet.MimeMessage.writeTo(MimeMessage.java:1913)
	at com.sun.mail.smtp.SMTPTransport.sendMessage(SMTPTransport.java:1315)
	at hudson.plugins.emailext.ExtendedEmailPublisher.sendMail(ExtendedEmailPublisher.java:524)
	at hudson.plugins.emailext.ExtendedEmailPublisher._perform(ExtendedEmailPublisher.java:444)
	at hudson.plugins.emailext.ExtendedEmailPublisher.perform(ExtendedEmailPublisher.java:354)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:814)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:763)
	at hudson.model.Build$BuildExecution.cleanUp(Build.java:189)
	at hudson.model.Run.execute(Run.java:1943)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:44)
	at hudson.model.ResourceController.execute(ResourceController.java:101)
	at hudson.model.Executor.run(Executor.java:442)
```

Googling the error message, it appears that JavaMail uses the thread's context class loader to load a configuration file mapping MIME types to Java classes. When the plugin is detached, the thread's context class loader is Jenkins core's class loader, which does not have these configuration files. I can work around the problem on the plugin side by wrapping calls to `Transport#send` in:

```
Thread t = Thread.currentThread();
ClassLoader orig = t.getContextClassLoader();
t.setContextClassLoader(getClass().getClassLoader());
try {
    […]
} finally {
    t.setContextClassLoader(orig);
}
```

I will think about how best to deal with this.

**Comment by basil** [comment]
As of jenkinsci/javax-mail-api-plugin#3 and jenkinsci/javax-mail-api-plugin#4 I have successfully completed end-to-end testing with both JavaMail and Jakarta Mail.

**Comment by basil** [comment]
If anyone wants to try out the Jakarta Mail version, here are the links:

Jenkins core: https://ci.jenkins.io/job/Core/job/jenkins/job/PR-6165/12/artifact/org/jenkins-ci/main/jenkins-war/2.329-rc31978.fa_f50a_a_d3e46/jenkins-war-2.329-rc31978.fa_f50a_a_d3e46.war
`mailer`: https://ci.jenkins.io/job/Plugins/job/mailer-plugin/job/PR-144/4/artifact/org/jenkins-ci/plugins/mailer/403.v4da_0b_7596fa_6/mailer-403.v4da_0b_7596fa_6.hpi
`email-ext`: https://ci.jenkins.io/job/Plugins/job/email-ext-plugin/job/PR-337/4/artifact/org/jenkins-ci/plugins/email-ext/2.87-rc1567.e14b_09c45d67/email-ext-2.87-rc1567.e14b_09c45d67.hpi
`workflow-basic-steps`: https://ci.jenkins.io/job/Plugins/job/workflow-basic-steps-plugin/job/PR-168/3/artifact/org/jenkins-ci/plugins/workflow/workflow-basic-steps/2.25-rc921.8e2290b_665f0/workflow-basic-steps-2.25-rc921.8e2290b_665f0.hpi

**Comment by oleg-nenashev** [comment]
Do we plan to get all downstream patches in https://github.com/jenkinsci/jenkins/pull/6165#issuecomment-1008054733 released before Wednesday? Mailer is out, but Mail Ext and `Pipeline: Basic Steps` aren't. CC @batmat @jtnord @rsandell @car-roll @bitwiseman  for the Pipeline plugins since they have permissions

**Comment by basil** [comment]
> Do we plan to get all downstream patches in [#6165 (comment)](https://github.com/jenkinsci/jenkins/pull/6165#issuecomment-1008054733) released before Wednesday?

No, this PR is not blocked on any downstream PRs. The downstream PRs can be merged and released at any time.

**Comment by timja** [comment]
> released before Wednesday

next weekly is security release so there's another week to go before it'll arrive to users anyway

**Comment by oleg-nenashev** [comment]
Thanks for patiently reminding me about that, Tim :)

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (daniel-beck) [inline core/src/main/resources/jenkins/split-plugins.txt:39]: https://groups.google.com/g/jenkinsci-advisories/c/wFaEbxbFSY0/m/KMza8OrfEAAJ + https://www.jenkins.io/security/scheduling/#changes-in-security-updates means that 2.329 will be released without this change, even if merged today.

```suggestion
javax-activation-api 2.329 1.2.0-2
javax-mail-api 2.329 1.6.2-5
```

File: core/src/main/resources/jenkins/split-plugins.txt
Comment (timja) [inline core/src/main/resources/jenkins/split-plugins.txt:39]: Weekly release accidentally happened before security release so this will be bumped a version
```suggestion
javax-activation-api 2.330 1.2.0-2
javax-mail-api 2.330 1.6.2-5
```

**Comment by jglick** [comment]
Might be handy to copy https://github.com/jenkinsci/ldap-plugin/blob/94fe2df805fe7daba02d432581aa2f625bc2e2cb/src/main/java/jenkins/security/plugins/ldap/SetContextClassLoader.java#L31-L47 to a utility package in core?

**Pull request merged**
